src/snowflake/connector/s3_storage_client.py

```diff
@@ -87,84 +92,134 @@ def __init__(
+
+        Does not support sorting by values in case the keys are the same, don't send
+        in duplicate keys!
+        """
         res = []
+        low_key_dict = {k.lower(): v for k, v in headers.items()}
+        sorted_headers = sorted(low_key_dict.keys())
+        _res = [(k, low_key_dict[k]) for k in sorted_headers]
 
-        for i in range(len(_res)):
-            k, v = _res[i]
+        for k, v in _res:
             # if value is a list, convert to string delimited by comma
             if isinstance(v, list):
                 v = ",".join(v)
             # if multiline header, replace withs space
             k = k.replace("\n", " ")
-            res.append(k.rstrip() + ":" + v.lstrip())
+            res.append(k.strip() + ":" + RE_MULTIPLE_SPACES.sub(" ", v.strip()))
 
         ans = "\n".join(res)
         if ans:
-            ans = ans + "\n"
+            ans += "\n"
 
-        return ans
+        return ans, ";".join(sorted_headers)
 
     @staticmethod
-    def _construct_string_to_sign(
+    def _construct_canonical_request_and_signed_headers(
         verb: str,
-        canonicalized_element: str,
-        canonicalized_headers: str,
-        amzdate: str,
-        content_md5: str = "",
-        content_type: str = "",
-    ) -> bytes:
-        res = verb + "\n" + content_md5 + "\n" + content_type + "\n"
-        res += amzdate + "\n" + canonicalized_headers + canonicalized_element
-        return res.encode("UTF-8")
+        canonical_uri_parameter: str,
+        query_parts: Dict[str, str],
+        canonical_headers: Optional[Dict[str, Union[str, List[str]]]] = None,
+        payload_hash: str = "",
+    ) -> Tuple[str, str]:
+        # Build canonical query string
+        # Note: this doesn't support sorting by values in case the same key is given
+        #  more than once, so don't do that!
+        canonical_query_string = "&".join(
+            "=".join([k, v]) for k, v in sorted(query_parts.items(), key=itemgetter(0))
+        )
+        (
+            canonical_headers,
+            signed_headers,
+        ) = SnowflakeS3RestClient._construct_canonicalized_and_signed_headers(
+            canonical_headers
+        )
+
+        return (
+            "\n".join(
+                [
+                    verb,
+                    canonical_uri_parameter or "/",
+                    canonical_query_string,
+                    canonical_headers,
+                    signed_headers,
+                    payload_hash,
+                ]
+            ),
+            signed_headers,
+        )
 
     @staticmethod
-    def _has_expired_token(response: requests.Response) -> bool:
+    def _construct_string_to_sign(
+        region_name: str,
+        service_name: str,
+        amzdate: str,
+        canonical_request_hash: bytes,
+    ) -> Tuple[str, str]:
+        short_amzdate = amzdate[:8]
+        scope = f"{short_amzdate}/{region_name}/{service_name}/aws4_request"
```

Comment:
What's short_amzdate for ? And can you share how it's different from amzdate?

Reply:
So from https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html As the first step you generate a canonical request and hash that to generate a string to sign, which we then hash and and include as a header to our HTTP request.

I think your question is best answered by examples:
Example canonical request:
```
GET
/
Action=ListUsers&Version=2010-05-08
content-type:application/x-www-form-urlencoded; charset=utf-8
host:iam.amazonaws.com
x-amz-date:20150830T123600Z

content-type;host;x-amz-date
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Assume which when hashed gives: `f536975d06c0309214f805bb90ccff089219ecd68b2577efef23edd43b7e1a59`

Subsequently derived string to sign:
```
AWS4-HMAC-SHA256
20150830T123600Z
20150830/us-east-1/iam/aws4_request
f536975d06c0309214f805bb90ccff089219ecd68b2577efef23edd43b7e1a59
```

Now see the blueprint for the string to sign:
```
StringToSign =
    Algorithm + \n +
    RequestDateTime + \n +
    CredentialScope + \n +
    HashedCanonicalRequest
```

Now notice that `CredentialScope` starts with a date in the example. Which is really just the first 8 characters of the `RequestDateTime`.

Reply:
Thanks for the explanation. Instead of hardcoding `[8:]` can we use an native datetime API and pass in the amzdate to it that returns the date part only? This change would then be needed at all instances where `short_amzdate` is derived in a similar manner.

Reply:
I believe `short_amzdate = amzdate[:8]` is more efficient than deriving another string from `datetime`.
```pycon
>>> import timeit
>>> import datetime
>>> t = datetime.datetime.utcnow()
>>> amzdate = t.strftime("%Y%m%dT%H%M%SZ")
>>> timeit.timeit("amzdate[:8]", globals=locals())
0.08361914503620937
>>> timeit.timeit("t.strftime('%Y%m%d')", globals=locals())
1.7320209199679084
```
I do see that I do `amzdate[:8]` twice tho, I can get rid of that second one 👍